### PR TITLE
add iter all function to typed-store

### PR DESCRIPTION
**Pull request by lanvidr**
This is needed for https://github.com/mystenlabs/narwhal/issues/115

### CODE REVIEW

File: crates/typed-store/src/lib.rs [hunk 1]
```diff
@@ -38,6 +38,7 @@ pub enum StoreCommand<Key, Value> {
     Read(Key, oneshot::Sender<StoreResult<Option<Value>>>),
     ReadAll(Vec<Key>, oneshot::Sender<StoreResult<Vec<Option<Value>>>>),
     NotifyRead(Key, oneshot::Sender<StoreResult<Option<Value>>>),
+    Iter(oneshot::Sender<HashMap<Key, Value>>),
```
Comment (gdanezis) [inline crates/typed-store/src/lib.rs:41]: The typed store iterator usually returns items in order of keys but a `HashMap` does not. Would it be better to use `BTreeMap` or even a Vec to preserve that order?

**Review by gdanezis — APPROVED** [review]
Nice one -- maybe preserve the order?

File: crates/typed-store/src/lib.rs (hunk 1 above)
Comment (lanvidr) [reply crates/typed-store/src/lib.rs:41]: For the use case I have, order doesn't matter at all, but this is a good point.

File: crates/typed-store/src/lib.rs [hunk 2]
```diff
@@ -120,6 +121,10 @@ where
                                 .push_back(sender)
                         }
                     }
+                    StoreCommand::Iter(sender) => {
+                        let response = keyed_db.iter().collect();
```
Comment (mystenmark) [inline crates/typed-store/src/lib.rs:125]: Wait a second - doesn't this load the whole table into memory?

File: crates/typed-store/src/lib.rs [hunk 3]
```diff
@@ -225,4 +230,14 @@ where
             .await
             .expect("Failed to receive reply to NotifyRead command from store")
     }
+
+    pub async fn iter(&self) -> HashMap<Key, Value> {
```
Comment (mystenmark) [inline crates/typed-store/src/lib.rs:234]: shouldn't this return `impl Iterator<Item = (Key, Value)>` - as is this implies loading the whole table into memory.

Comment (lanvidr) [reply crates/typed-store/src/lib.rs:234]: this was addressed offline

**Pull request merged**